### DISCUSSION TITLE
GitHub Branch Protection Rules Queries Stopped Working

### DISCUSSION THREAD

**Discussion by rgajason** [profusion/sgqlc#216] [Q&A] [answered]
In the past I was able to retrieve a GitHub repository's branch protection rules like so:

```python
op = Operation(schema.Query)
repo = op.repository(owner="myorg", name="myrepo")
repo.id()
repo.name()
repo.description()
repo.name_with_owner()
repo.branch_protection_rules(first=10)
data = gql.endpoint(op)
```

...where `schema` is loaded from the file created by `sgqlc-codegen`:

```shell
python3 -m sgqlc.introspection \
    --exclude-deprecated \
    --exclude-description \
    -H "Authorization: bearer ${GITHUB_PAT}" \
    https://api.github.com/graphql \
    github_schema.json
sgqlc-codegen schema github_schema.json github_schema.py
```

Recently, however, when I attempt to add BPRs to the repo query I get the following error:

```python
>>> repo.branch_protection_rules(first=10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/path/to/my/virtualenv/lib/python3.10/site-packages/sgqlc/operation/__init__.py", line 1811, in __call__
    raise ValueError(
ValueError: branch_protection_rules already have a selection branchProtectionRules {
  pageInfo {
    endCursor
    hasNextPage
    hasPreviousPage
    startCursor
  }
  edges {
    cursor
  }
  nodes {
    id
    allowsDeletions
    allowsForcePushes
    blocksCreations
    databaseId
    dismissesStaleReviews
    isAdminEnforced
    lockAllowsFetchAndMerge
    lockBranch
    pattern
    requireLastPushApproval
    requiredApprovingReviewCount
    requiredStatusCheckContexts
    requiresApprovingReviews
    requiresCodeOwnerReviews
    requiresCommitSignatures
    requiresConversationResolution
    requiresLinearHistory
    requiresStatusChecks
    requiresStrictStatusChecks
    restrictsPushes
    restrictsReviewDismissals
  }
  totalCount
}. Maybe use __alias__ as param?
```

I'm using release 16.0 of `sgqlc` and pulled the GH GQL schema down today to ensure I have latest.

Thoughts?  Thanks!

**Comment by rgajason** [comment, marked as answer]
EDIT :: Real answer is in the comment thread below.  TL;DR: issue with CLI vs script.



In true fashion, I was able to get it sorta working in my very next test...

This fails:

```
repo = op.repository(owner="myorg", name="myrepo")
repo.branch_protection_rules(first=10)
```

But this works:

```
bprs = op.repository(owner="myorg", name="myrepo").branch_protection_rules(first=10)
```

This also fails:

```
bprs = op.repository(owner="myorg", name="myrepo").branch_protection_rules(first=10)
repo = op.repository(owner="myorg", name="myrepo")
```

The `brps =` assignment works but the following `repo =` assignment fails with:

```
ValueError: repository already have a selection repository(...)
```

Would be interested in thoughts on how I should be doing these queries so that I don't have to create multiple operations.

**Reply by barbieri** [reply]
could you try to pin-point the version that caused your issue? you may use `git bisect` to help your work (ask it to execute your test).

The code looks correct. The only thing that comes to my mind is the auto-selection work, but that would. only be the case if you're printing or somehow serializing the query... Your first example is running within the console, where print automatically happens, but does it happen in actual code scripts?

My only comment would be to avoid recreating queries with parameters, you should be saving them:

```py
bprs = op.repository(owner="myorg", name="myrepo").branch_protection_rules(first=10)
repo = op.repository(owner="myorg", name="myrepo")
```

Should be:

```py
repo = op.repository(owner="myorg", name="myrepo")
bprs = repo.branch_protection_rules(first=10)
```

But in any case, it's likely a bug because auto-selections shouldn't be generated if there is already a selection... and they shouldn't be saved internally unless explicitly with `__fields__()`

**Reply by barbieri** [reply]
looking at git history, maybe https://github.com/profusion/sgqlc/commit/4735ccaeb9e2f1127683ff74e469332a740ea99d is the failure... try with that commit and the previous one (https://github.com/profusion/sgqlc/commit/4afc6842353ea9f08d6fee38af3448e1d48f547d or https://github.com/profusion/sgqlc/commit/47968a58928cd3057cec868fc8e4e6c7da37375e)

**Reply by rgajason** [reply]
Thanks for the response - I wanted to acknowledge and let you know that I will follow up after the weekend.

**Reply by rgajason** [reply]
Same issue exists in 4afc684 and 47968a5 (assuming I'm doing this right):

```shell
virtualenv venv
source venv/bin/activate
pip install sgqlc @ git+ssh://github.com/profusion/sgqlc.git@4afc6842353ea9f08d6fee38af3448e1d48f547d
pip install requests
python
```

Then, in the Python CLI session:

```python
import requests
from sgqlc.endpoint.requests import RequestsEndpoint
from sgqlc.operation import Operation
from github_schema import github_schema as schema
from rich.pretty import pprint

endpoint = RequestsEndpoint(
    url="https://api.github.com/graphql",
    base_headers={
        "Authorization": "bearer my_pat_here",
        "Accept": "application/vnd.github.v4.idl",
    },
    session=requests.Session(),
)
op = Operation(schema.Query)
repo = op.repository(owner="my_org_here", name="my_repo_name_here")
repo.branch_protection_rules(first=10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/path/to/venv/lib/python3.10/site-packages/sgqlc/operation/__init__.py", line 1708, in __call__
    raise ValueError(
ValueError: branch_protection_rules already have a selection branchProtectionRules {
  pageInfo {
    endCursor
    hasNextPage
    hasPreviousPage
    startCursor
  }
  edges {
    cursor
  }
  nodes {
    id
    allowsDeletions
    allowsForcePushes
    blocksCreations
    databaseId
    dismissesStaleReviews
    isAdminEnforced
    lockAllowsFetchAndMerge
    lockBranch
    pattern
    requireLastPushApproval
    requiredApprovingReviewCount
    requiredStatusCheckContexts
    requiresApprovingReviews
    requiresCodeOwnerReviews
    requiresCommitSignatures
    requiresConversationResolution
    requiresLinearHistory
    requiresStatusChecks
    requiresStrictStatusChecks
    restrictsPushes
    restrictsReviewDismissals
  }
  totalCount
}. Maybe use __alias__ as param?
```

(repeat with clean virtual environment and `sgqlc` from 47968a58928cd3057cec868fc8e4e6c7da37375e)

I'll have to do a bit of research on `git bisect`.

**Reply by rgajason** [reply]
Okay, sooo... take that code, put it in a file, set the shebang, make it executable, run it - works.

In other words - if I type the code into the Python CLI I get the error.
If I put the exact same code into a file and execute as a script, it works.

ლ(ಠ益ಠლ)

**Reply by barbieri** [reply]
yeah, it's still a bug since the auto-selection shouldn't be persisted... but ok, at least can get you going :-)
